File: frontend/src/api/datasource.ts
import { api } from './request';

export interface Datasource {
  id: number;
  datasource_code: string;
  datasource_name: string;
  db_type: string;
  host: string;
  port: number;
  database_name?: string;
  schema_name?: string;
  username: string;
  charset?: string;
  connect_timeout_seconds?: number;
  status: string;
  last_test_status?: string;
  last_test_message?: string;
  last_test_at?: string;
  remark?: string;
  created_at: string;
  updated_at: string;
}

export interface DatasourceCreate {
  datasource_name: string;
  db_type: string;
  host: string;
  port: number;
  database_name?: string;
  schema_name?: string;
  username: string;
  password: string;
  charset?: string;
  connect_timeout_seconds?: number;
  status?: string;
  remark?: string;
}

export interface TestConnectionReq {
  db_type: string;
  host: string;
  port: number;
  database_name?: string;
  schema_name?: string;
  username: string;
  password: string;
  charset?: string;
  connect_timeout_seconds?: number;
}

export const listDatasources = (params?: Record<string, unknown>) =>
  api.get<Datasource[]>('/datasource', { params });

export const countDatasources = (params?: Record<string, unknown>) =>
  api.get<{ total: number }>('/datasource/count', { params });

export const getDatasource = (id: number) =>
  api.get<Datasource>(`/datasource/${id}`);

export const createDatasource = (data: DatasourceCreate) =>
  api.post<Datasource>('/datasource', data);

export const updateDatasource = (id: number, data: Partial<DatasourceCreate>) =>
  api.put<Datasource>(`/datasource/${id}`, data);

export const deleteDatasource = (id: number) =>
  api.delete(`/datasource/${id}`);

export const testConnection = (data: TestConnectionReq) =>
  api.post<{ success: boolean; message: string }>('/datasource/test-connection', data);

export const testExistingDatasource = (id: number) =>
  api.post<{ success: boolean; message: string }>(`/datasource/${id}/test`);

File: backend/app/schemas/datasource.py
"""Pydantic schemas for datasource endpoints."""

from __future__ import annotations
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class DatasourceCreate(BaseModel):
    datasource_name: str = Field(..., max_length=128)
    db_type: str = Field(..., pattern=r"^(mysql|postgresql|sqlserver|oracle|dm|kingbase|sqlite)$")
    host: str = Field("", max_length=255)
    port: int = Field(0, ge=0, le=65535)
    database_name: Optional[str] = None
    schema_name: Optional[str] = None
    username: str = Field(..., max_length=128)
    password: str = Field("", min_length=0)
    charset: Optional[str] = "utf8"
    connect_timeout_seconds: Optional[int] = 10
    status: str = "enabled"
    remark: Optional[str] = None


class DatasourceUpdate(BaseModel):
    datasource_name: Optional[str] = None
    db_type: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    database_name: Optional[str] = None
    schema_name: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None
    charset: Optional[str] = None
    connect_timeout_seconds: Optional[int] = None
    status: Optional[str] = None
    remark: Optional[str] = None


class DatasourceOut(BaseModel):
    id: int
    datasource_code: str
    datasource_name: str
    db_type: str
    host: str
    port: int
    database_name: Optional[str] = None
    schema_name: Optional[str] = None
    username: str
    charset: Optional[str] = None
    connect_timeout_seconds: Optional[int] = None
    status: str
    last_test_status: Optional[str] = None
    last_test_message: Optional[str] = None
    last_test_at: Optional[datetime] = None
    remark: Optional[str] = None
    created_by: str
    created_at: datetime
    updated_by: str
    updated_at: datetime

    model_config = {"from_attributes": True}


class TestConnectionRequest(BaseModel):
    db_type: str
    host: str
    port: int
    database_name: Optional[str] = None
    schema_name: Optional[str] = None
    username: str
    password: str
    charset: Optional[str] = "utf8"
    connect_timeout_seconds: Optional[int] = 10


class TestConnectionResponse(BaseModel):
    success: bool
    message: str
